ghcide/src/Development/IDE/Plugin/TypeLenses.hs

```diff
@@ -91,28 +101,35 @@ instance Pretty Log where
     LogShake msg -> pretty msg
 
 
-typeLensCommandId :: T.Text
+typeLensCommandId :: IsString s => s
 typeLensCommandId = "typesignature.add"
 
 descriptor :: Recorder (WithPriority Log) -> PluginId -> PluginDescriptor IdeState
 descriptor recorder plId =
   (defaultPluginDescriptor plId desc)
     { pluginHandlers = mkPluginHandler SMethod_TextDocumentCodeLens codeLensProvider
                     <> mkResolveHandler SMethod_CodeLensResolve codeLensResolveProvider
-    , pluginCommands = [PluginCommand (CommandId typeLensCommandId) "adds a signature" commandHandler]
-    , pluginRules = rules recorder
+                    <> mkPluginHandler SMethod_TextDocumentInlayHint localBindingInlayHints
+                    <> mkPluginHandler SMethod_TextDocumentCodeLens localBindingCodeLens
+    , pluginCommands = [PluginCommand typeLensCommandId "adds a signature" commandHandler]
+    , pluginRules = globalBindingRules recorder *> localBindingRules recorder
     , pluginConfigDescriptor = defaultConfigDescriptor {configCustomConfig = mkCustomConfig properties}
     }
   where
-    desc = "Provides code lenses type signatures"
+    desc = "Provides type signatures through code lenses and inlay hints"
 
-properties :: Properties '[ 'PropertyKey "mode" (TEnum Mode)]
+properties :: Properties
+  '[ 'PropertyKey "localBindingInlayHintOn" 'TBoolean,
+     'PropertyKey "mode" ('TEnum Mode)]
 properties = emptyProperties
   & defineEnumProperty #mode "Control how type lenses are shown"
     [ (Always, "Always displays type lenses of global bindings")
     , (Exported, "Only display type lenses of exported global bindings")
     , (Diagnostics, "Follows error messages produced by GHC about missing signatures")
     ] Always
+  & defineBooleanProperty #localBindingInlayHintOn
+    "Display inlay hints of local bindings"
```

Comment:
this needs a better explanation for documentation

Reply:
this may also be clearer as a three-way enum called `localBindingHintsMode` with the options `NoLocalHints`, `UseInlayHints`, `UseCodeLenses`.

happy to let others bikeshed this, though.